### DISCUSSION TITLE
Options for greater USD scene modifications on export?

### DISCUSSION THREAD

**Discussion by Rhinoceroscity** [Autodesk/maya-usd#1890] [Q&A]
Hey everyone, we're looking at more options for transforming data on export from Maya into USD, and wanted to get some perspectives if anyone else has tried this. We've already investigated export chasers, and while useful, their utility is somewhat limited by the fact that it can really only create and set basic attributes on existing prims. They don't seem to have the ability to create new prims altogether, or add layers to the scene, and we still aren't sure whether we can store more complex data (such as arrays of compound attributes, or nested compound attributes) in a way that isn't fairly clunky (such as creating new array attributes for each member of a structure).

Our thought right now is that we'll have an export chaser that provides all the basic information necessary on each prim, and then afterwards, in a post export step (meaning a separate script or executable), load the newly saved USD file and re-process it to setup new layers and prims and such based on the data provided on each node, and then save it back out again.

Has anyone tried something like this, or have a perspective on whether this is a good path to take? One potential pitfall being on re-import back into Maya or another package, we would need to have metadata that describes how to reconstruct the original node based on what was originally exported directly from Maya before our post modifications.

Any advice or info would be appreciated!

**Comment by dgovil** [comment]
Export chasers can currently already create new prims. We make use of that in our pipeline currently.
You should have full access to the current layer being exported.

We override and implement the `ExportDefault` method of the chaser and create our prims there as needed.

**Reply by Rhinoceroscity** [reply]
Huh, that's interesting, the comment for chasers in the header file reads as such:

/// Chasers need to be very careful as to not modify the structure of the usd
/// file.  This should ideally be used to make small changes or to add
/// attributes in a non-destructive way.

Is creating a new prim not considered modification of the structure of the USD file?

**Reply by dgovil** [reply]
Someone else may correct me, but I think that as long as changes are additive not subtractive its fine. e.g as long as you're not moving something out from under the rest of the exporter code, otherwise it'll depend on it being there and potentially fail.

**Reply by Rhinoceroscity** [reply]
Okay awesome, that is good to know, thank you very much for your input!

**Comment by neilrubie** [comment]
We certainly do this. There are many post-export modifications we need to be done for our specific use cases that are just not possible with the exporter as is. We typically export a usd cache to a tmp location, then open it up and traverse as necessary making changes all over the place. In such cases we don't worry about the round trip back into Maya since this cache isn't necessarily meant to make sense on it's own.  It is often just a single piece of a larger usd composition that only makes sense as a whole.

**Reply by mdubrois** [reply]
We are in the same case, export tmp (with a bit of chaser) then we open and modify the USD according to the needs.
